features/filesystem/bd/BufferedBlockDevice.cpp

```diff
@@ -0,0 +1,229 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "BufferedBlockDevice.h"
+#include "mbed_assert.h"
+#include <algorithm>
+#include <string.h>
+
+static inline uint32_t align_down(bd_size_t val, bd_size_t size)
+{
+    return val / size * size;
+}
+
+BufferedBlockDevice::BufferedBlockDevice(BlockDevice *bd)
+    : _bd(bd), _bd_program_size(0), _curr_aligned_addr(0), _flushed(true), _cache(0)
+{
+}
+
+BufferedBlockDevice::~BufferedBlockDevice()
+{
+    deinit();
+    if (_cache) {
+        delete[] _cache;
+    }
+}
+
+int BufferedBlockDevice::init()
+{
+    int err = _bd->init();
+    if (err) {
+        return err;
+    }
+
+    _bd_program_size = _bd->get_program_size();
+
+    if (!_cache) {
+        _cache = new uint8_t[_bd_program_size];
```

Comment:
If this alloc is being deferred to the error-returning `init`, does it make sense to use `(nothrow)` and return an error on memory failure?

Reply:
Reason for this being allocated in init is the fact that it relies on the underlying BD's `get_program_size`, whose value may only be valid after init. Code originally included an assertion for allocation failure, but then was told by @geky that it was redundant, as mbed-os automatically asserts on allocation failures. Is this not correct?

Reply:
That's correct - it just seems slightly odd to throw given that all other errors in the function return an error code.

On the other hand, it's not that you really want dynamic allocation with check - if it weren't for that `get_program_size` you'd be doing a throwing `new` in the constructor, right? So maybe it's just that, deferred.

I'm generally in favour of throwing for unexpected errors, rather than returning an error no-one checks. So fine.

Reply:
I try to distinguish "fatal" errors (like not being able to allocate at init) from "non fatal" ones (errors in read/write for wrong addresses etc.). At least my code (like StorageLite) checks for error returned by the latter. Maybe I should return the assert for allocation failure just for visibility sake.

Reply:
I think the question is "would an application want to recover from an OOM in the buffered block device?"